### PR TITLE
[CORDA-2390] - Add whitelists and custom serializers from cordapps to serialization …

### PR DESCRIPTION
**Pull request by dimosr**
## Changes
This introduces 2 main changes:
* add the whitelists from the cordapps (loaded from the cordapps folder during node initialisation) to the serialization scheme
* enhance the creation of attachments class loader context, so that it scans the classpath for available custom serializers/whitelists. This will be used in cases, like transaction verification, where there could be serializers in the attachments that have been added (or fetched from peers) post node initialisation

Notes:
* The custom serializers for the SIMM valuation demo were moved together with the contracts (they were in the folder of the flows before). The reason is that attachments are automatically added to a transaction, when there are missing states/contracts, by auto-discovering attachments that contain them. As a result, in order to be included in this process, custom serializers must reside in the same cordapp with the contracts/states.
* The `TransactionBuilder` contains a small patch, provided by @tudor-malene (to cover some cases of transactions with input states)
* As mentioned in the comments, the `CurrencySerializers` temporarily overrides the `threadContextClassLoader`. Arguably, this could be done at a higher-level to capture all cases (i.e. when triggering an RPC call), but that might have bigger consequences and avoided it for now, since this change is quite big. Also, as described, the behaviour of this library is rather peculiar and I don't anticipate that being made by other 3rd party libraries in other demos currently.
* Florian mentioned that there might be some demos using Spring, so I'll try to find those as well and ensure they work, or adjust them in the same way as I did in the webserver. I want to decouple this from this, so I'll open a separate PR.

## Testing
* Run the SIMM valuation demo and it works fine now
* (In the meanwhile), I want to also run some other demo (perhaps trading) to make sure there are no regressions

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt [hunk 1]
```diff
@@ -148,12 +150,20 @@ internal object AttachmentsClassLoaderBuilder {
     }
 
     fun <T> withAttachmentsClassloaderContext(attachments: List<Attachment>, block: (ClassLoader) -> T): T {
-
         // Create classloader from the attachments.
         val transactionClassLoader = AttachmentsClassLoaderBuilder.build(attachments)
 
+        val serializers = loadClassesImplementing(transactionClassLoader, SerializationCustomSerializer::class.java)
+        val whitelistedClasses = ServiceLoader.load(SerializationWhitelist::class.java, transactionClassLoader)
```
Comment (tudor-malene) [inline core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: These need to be cached similar to the classloader.

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 2]
```diff
@@ -337,15 +338,15 @@ open class TransactionBuilder @JvmOverloads constructor(
...
         val forcedAttachmentId = explicitContractAttachment ?: hashAttachments.singleOrNull()?.id
 
-        fun selectAttachment() = selectAttachmentThatSatisfiesConstraints(
-                false,
-                contractClassName,
-                inputsAndOutputs.filterNot { it.constraint in automaticConstraints },
-                inputStateRefs,
-                services)
+        // States that have an explicit constraint
+        val explicitConstraintStates = inputsAndOutputs.filterNot { it.constraint in automaticConstraints }
```
Comment (tudor-malene) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: Please revert the changes to this file.
It was most likely a red herring.

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt [hunk 3]
```diff
@@ -43,11 +46,12 @@ interface SerializerFactoryFactory {
 @KeepForDJVM
 abstract class AbstractAMQPSerializationScheme(
```
Comment (dimosr) [inline serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt]: The `serializationWhitelists` below scans the class path and adds everything found on top of the default whitelist. I am wondering whether we should remove that now.

P.S.: I tried removing and re-running the demo and it worked fine, but not sure how it's used anywhere else and whether there's any concern on backwards compatibility.

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt (hunk 1 above)
Comment (dimosr) [reply core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: Good catch. 
Noticed the cache is keyed based on the list of attachments' hashes. I was thinking of switching it to a `Set<SecureHash>` and remove the sorting. What do you think ?

Comment (tudor-malene) [reply core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: The AttachmentId is a SecureHash. 
Set or sorted list is pretty much the same . Your call.

File: samples/simm-valuation-demo/contracts-states/src/main/kotlin/net/corda/vega/plugin/customserializers/CurrencySerializer.kt [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+    /**
+     * The initialization of [Currency] uses the classpath to identify needed resources.
+     * However, it gives priority to the classloader in the thread context over the one this class was loaded with.
+     * See: [com.opengamma.strata.collect.io.ResourceLocator.classLoader]
+     *
+     * This is the reason we temporarily override the class loader in the thread context here, with the classloader of this
+     * class, which is guaranteed to contain everything in the 3rd party library's classpath.
+     */
+    private fun withCurrentClassLoader(serializationFunction: () -> Currency): Currency {
```
Comment (tudor-malene) [inline samples/simm-valuation-demo/contracts-states/src/main/kotlin/net/corda/vega/plugin/customserializers/CurrencySerializer.kt]: This is required because it uses some resource from the `contextClassLoader `?

**Comment by tudor-malene** [comment]
@shamsasari, can you take a look at this from the pov of the API ?
I'll try to focus on the functional bits.

File: samples/simm-valuation-demo/contracts-states/src/main/kotlin/net/corda/vega/plugin/customserializers/CurrencySerializer.kt (hunk 4 above)
Comment (dimosr) [reply samples/simm-valuation-demo/contracts-states/src/main/kotlin/net/corda/vega/plugin/customserializers/CurrencySerializer.kt]: Yes, it tries to load some resources (`Currency.ini`) files, but gives priority to the threadContextClassLoader, instead of the current classloader for some reason.

File: client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt [hunk 5]
```diff
@@ -240,20 +244,22 @@ open class CordaRPCClientConfiguration @JvmOverloads constructor(
...
  * The client will attempt to connect to a live server by trying each address in the list. If the servers are not in
  * HA mode, the client will round-robin from the beginning of the list and try all servers.
  */
 class CordaRPCClient private constructor(
         private val hostAndPort: NetworkHostAndPort,
         private val configuration: CordaRPCClientConfiguration = CordaRPCClientConfiguration.DEFAULT,
         private val sslConfiguration: ClientRpcSslOptions? = null,
-        private val classLoader: ClassLoader? = null,
-        private val haAddressPool: List<NetworkHostAndPort> = emptyList()
+        private val haAddressPool: List<NetworkHostAndPort> = emptyList(),
```
Comment (tudor-malene) [inline client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt]: What's the reason of switching the order here?

File: common/classloading/build.gradle [hunk 6]
```diff
@@ -0,0 +1,23 @@
+apply plugin: 'kotlin'
```
Comment (tudor-malene) [inline common/classloading/build.gradle]: I'm not sue if these utility functions deserve their own module.
@shamsasari , wdyt?

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt [hunk 7]
```diff
@@ -130,30 +131,49 @@ class AttachmentsClassLoader(attachments: List<Attachment>, parent: ClassLoader
...
  */
 @VisibleForTesting
 internal object AttachmentsClassLoaderBuilder {
 
-    private const val ATTACHMENT_CLASSLOADER_CACHE_SIZE = 1000
+    private const val CACHE_SIZE = 1000
 
     // This runs in the DJVM so it can't use caffeine.
-    private val cache: MutableMap<List<SecureHash>, AttachmentsClassLoader> = createSimpleCache<List<SecureHash>, AttachmentsClassLoader>(ATTACHMENT_CLASSLOADER_CACHE_SIZE)
+    private val classloaderCache: MutableMap<Set<SecureHash>, AttachmentsClassLoader> = createSimpleCache<Set<SecureHash>, AttachmentsClassLoader>(CACHE_SIZE)
```
Comment (tudor-malene) [inline core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: Why not create a structure with the Classloader, the serializers and the whitelist, and just cache it once.
@r3domfox , is the `SerializationContext` safe for caching?

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 8]
```diff
@@ -17,6 +17,7 @@ import net.corda.core.node.services.KeyManagementService
 import net.corda.core.serialization.SerializationContext
```
Comment (tudor-malene) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: Please revert changes to this class too.

File: docs/source/changelog.rst [hunk 9]
```diff
@@ -52,6 +52,9 @@ Version 4.0
 
 * Fixed a problem that was preventing `Cash.generateSpend` to be used more than once per transaction (https://github.com/corda/corda/issues/4110).
 
+* The experimental confidential-identities is now a separate CorDapp and must now be loaded onto the node alongside any CorDapp that needs it.
```
Comment (tudor-malene) [inline docs/source/changelog.rst]: What's up with this?

File: samples/simm-valuation-demo/contracts-states/src/main/kotlin/net/corda/vega/plugin/customserializers/CurrencySerializer.kt (hunk 4 above)
Comment (tudor-malene) [reply samples/simm-valuation-demo/contracts-states/src/main/kotlin/net/corda/vega/plugin/customserializers/CurrencySerializer.kt]: I think this will be a problem in general.
I wonder if we should do the following:
1) On the RPC threads, set the `currentThread. contextClassLoader` to the cordapp classloader
2) On the flow threads we already do 
3) In AttachmentsClassloader.withAttachmentsClassloader, also do that.
  - before: read the contextClassLoader
  - set it to the transactionClassloader
  - finally: set it to the previous value

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt (hunk 3 above)
Comment (tudor-malene) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt]: You mean removing the default whitelist?

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt [hunk 10]
```diff
@@ -146,6 +150,15 @@ abstract class AbstractAMQPSerializationScheme(
                 factory.registerExternal(CorDappCustomSerializer(customSerializer, factory))
             }
         } else {
+            // This step is registering custom serializers, which have been added after node initialisation (i.e. via attachments during transaction verification).
+            // Note: the order between the registration of customSerializers and cordappCustomSerializers must be preserved as-is. The reason is the following:
+            // Currently, the serialization infrastructure does not support multiple versions of a class (the first one that is registered dominates).
+            // As a result, when inside a context with attachments class loader, we prioritize serializers loaded on-demand from attachments to serializers that had been
```
Comment (tudor-malene) [inline serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt]: This sounds really hacky.
@r3domfox, is it possible to register the serializers but store them based on the classloader.

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt (hunk 7 above)
Comment (dimosr) [reply core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: regarding the first point, makes sense, I'll do it. 
If @r3domfox confirms the context is safe for caching, then I can also just cache this, I guess.

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt (hunk 3 above)
Comment (dimosr) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt]: No, keeping the default whitelist, but remove the whitelist that is created by scanning the classpath.

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt (hunk 8 above)
Comment (dimosr) [reply core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: urgh, for some reason they resurrected, prob. during the rebase. I'll do it again.

File: docs/source/changelog.rst (hunk 9 above)
Comment (dimosr) [reply docs/source/changelog.rst]: I had to rebase from master, so these are probably coming from previous commits :/

File: samples/simm-valuation-demo/contracts-states/src/main/kotlin/net/corda/vega/plugin/customserializers/CurrencySerializer.kt (hunk 4 above)
Comment (dimosr) [reply samples/simm-valuation-demo/contracts-states/src/main/kotlin/net/corda/vega/plugin/customserializers/CurrencySerializer.kt]: What kind of problem can it create ?

The way I see it, what this library is doing is non-sensible. It seems reasonable to me that the custom serializers are supposed to work in the Cordapp classpath by-design (so any resources needed should be available in the current classloader). However, this library considers the classloader in the thread context to be more relevant than what the current class used.

Am I missing smth ?

Comment (tudor-malene) [reply samples/simm-valuation-demo/contracts-states/src/main/kotlin/net/corda/vega/plugin/customserializers/CurrencySerializer.kt]: A problem with libraries trying to read resources from the contextClassLoader when they are initialised.
If we set it ourselves, we avoid this problem, instead of expecting people to do it in serializers.

File: docs/source/changelog.rst (hunk 9 above)
Comment (dimosr) [reply docs/source/changelog.rst]: This was a recent change from Shams to refactor confidential-identities in a separate Cordapp. Actually, I needed to do that now for SIMM valuation as well.

File: samples/simm-valuation-demo/contracts-states/src/main/kotlin/net/corda/vega/plugin/customserializers/CurrencySerializer.kt (hunk 4 above)
Comment (dimosr) [reply samples/simm-valuation-demo/contracts-states/src/main/kotlin/net/corda/vega/plugin/customserializers/CurrencySerializer.kt]: mm, I see. It makes sense, from a customer POV, it's really a much better CX. What I don't like is that it kind of encourages bad patterns on the client-side.  

I also can't see clearly if doing that in all these places can cause other side-effects, based on where this thread context is consulted. Still, all these places are supposed to run in the context of the Cordapps, so even in that case, it should not change anything dramatically.

Comment (tudor-malene) [reply samples/simm-valuation-demo/contracts-states/src/main/kotlin/net/corda/vega/plugin/customserializers/CurrencySerializer.kt]: Reading from the `contextClassLoader` could happen deep in some library the cordapp uses, outside the control of the client.
By default the `contextClassLoader` is set to the system classloader, so it will be a superset of that.

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt (hunk 7 above)
Comment (r3domfox) [reply core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: I suppose it would save a lookup against the threadlocal? The intention is that the context should be immutable, so it should be safe to cache it.

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt (hunk 10 above)
Comment (r3domfox) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt]: Not really, no: they're cached against their declared `typeNotation`, which is also used to identify types serialized with them on the wire, so when we receive a value serialized by a custom serializer in a remote node, we can look up the local custom serializer immediately by that cache key. It would not make sense to include the classloader as part of that key.

We could make this less hacky by making the override order explicit, e.g. by only trying to register those `cordappCustomSerializers` that were not overridden by `customSerializers`, so that if someone changed the code around for some reason things wouldn't start failing mysteriously.

File: samples/simm-valuation-demo/contracts-states/src/main/kotlin/net/corda/vega/plugin/customserializers/CurrencySerializer.kt (hunk 4 above)
Comment (dimosr) [reply samples/simm-valuation-demo/contracts-states/src/main/kotlin/net/corda/vega/plugin/customserializers/CurrencySerializer.kt]: Good, I'll do what you mentioned then, thanks.

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt (hunk 10 above)
Comment (tudor-malene) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt]: I wonder, instead of using the defaultFactory, can we make a copy of it and then register these classloader specific serializers on the copy. 
And maybe we can cache the factory

Comment (r3domfox) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt]: The factories _are_ cached - there is one stored in the cache in `AbstractAMQPSerializationSchema` for every distinct combination of whitelist, classloader, etc in the `SerializationContext`.

Comment (tudor-malene) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt]: Right, so would that be feasible then?
( Create factories on the fly by copying the default one , and customizing them based on the classloader)

Comment (r3domfox) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt]: The state of a factory is basically a collection of caches. If you took an existing factory and modified the classloader, you would have to rebuild the cached objects (since they might refer to classes that wouldn't be visible from the new classloader, which is a recipe for total chaos) - it's easier just to create a new factory with empty caches, and populate them as needed.

Comment (r3domfox) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt]: We are in fact creating factories on the fly, configuring them according to what the `SerializationContext` says they they should be doing - setting evolution rules, whitelist, classloader and registering custom serialisers - and then caching them.

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt [hunk 11]
```diff
@@ -34,7 +35,7 @@ class AttachmentsClassLoader(attachments: List<Attachment>, parent: ClassLoader
     init {
         val untrusted = attachments.mapNotNull { it as? ContractAttachment }.filterNot { isUploaderTrusted(it.uploader) }.map(ContractAttachment::id)
         if(untrusted.isNotEmpty()) {
-            throw MissingAttachmentsException(untrusted, "Attempting to load Contract Attachments downloaded from the network")
+            throw UntrustedAttachmentsException(untrusted)
```
Comment (tudor-malene) [inline core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: This change does not belong to this PR

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt (hunk 10 above)
Comment (tudor-malene) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/AMQPSerializationScheme.kt]: After chatting with @r3domfox , this solution would do for now, but it needs some more thinking in a future release.

**Comment by tudor-malene** [comment]
I've ran the test and I get an OutOfMemory in the "ClassGraph-worker-6", so it seems to be a problem there.

File: serialization/src/test/kotlin/net/corda/serialization/internal/CordaClassResolverTests.kt [hunk 12]
```diff
@@ -217,8 +216,8 @@ class CordaClassResolverTests {
         CordaClassResolver(emptyWhitelistContext).getRegistration(attachedClass)
     }
 
-    @Test(expected = MissingAttachmentsException::class)
-    fun `Attempt to load contract attachment with the incorrect uploader should fails with MissingAttachmentsException`() {
+    @Test(expected = UntrustedAttachmentsException::class)
```
Comment (tudor-malene) [inline serialization/src/test/kotlin/net/corda/serialization/internal/CordaClassResolverTests.kt]: This does not belong here either

File: docs/source/release-notes.rst [hunk 13]
```diff
@@ -287,7 +287,13 @@ The TLS_DHE_RSA_WITH_AES_128_GCM_SHA256 family of ciphers is retired from the li
 as it is a legacy cipher family not supported by all native SSL/TLS implementations. We anticipate that this
 will have no impact on any deployed configurations.
 
-Miscelleneous changes
+Confidential identities
++++++++++++++++++++++++
+
+If any of your CorDapps use the experimental confidential-identities module then it also needs to be loaded as a separate CorDapp jar. This
```
Comment (tudor-malene) [inline docs/source/release-notes.rst]: this too

File: docs/source/cordapp-build-systems.rst [hunk 14]
```diff
@@ -471,4 +474,36 @@ For a CorDapp that contains flows and/or services we specify the `workflow` tag:
             }
         }
 
-.. note:: It is possible, but *not recommended*, to include everything in a single CorDapp jar and use both the ``contract`` and ``workflow`` Gradle plugin tags.
\ No newline at end of file
+.. note:: It is possible, but *not recommended*, to include everything in a single CorDapp jar and use both the ``contract`` and ``workflow`` Gradle plugin tags.
```
Comment (tudor-malene) [inline docs/source/cordapp-build-systems.rst]: this too

Comment (dimosr) [reply docs/source/cordapp-build-systems.rst]: As discussed, all of these are because of the rebase & merge, they are currently all merged in master.

Let me know if it obstructs reviewing, I can force-push and keep only my changes in the PR.

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt (hunk 7 above)
Comment (dimosr) [reply core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: cool, I'll cache that instead then

File: docs/source/cordapp-build-systems.rst (hunk 14 above)
Comment (tudor-malene) [reply docs/source/cordapp-build-systems.rst]: Yes please. A PR should only contain relevant changes.

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/NotarySpec.kt [hunk 15]
```diff
@@ -19,7 +19,8 @@ data class NotarySpec(
         val validating: Boolean = true,
         val rpcUsers: List<User> = emptyList(),
         val verifierType: VerifierType = VerifierType.InMemory,
-        val cluster: ClusterSpec? = null
+        val cluster: ClusterSpec? = null,
+        val maximumHeapSize: String = "512m"
```
Comment (anthonykeenan) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/NotarySpec.kt]: Adding a variable to a data class causes an API breakage - if you create a seperate constructor that mimcs the old one, and you'll also need to create a copy method with defaults:

These are the lines from the build:
```
[11:45:49][Step 4/10] -  public <init>(net.corda.core.identity.CordaX500Name, boolean, java.util.List<net.corda.testing.node.User>, net.corda.testing.driver.VerifierType, net.corda.testing.node.ClusterSpec)
[11:45:49][Step 4/10] -  public final net.corda.testing.node.NotarySpec copy(net.corda.core.identity.CordaX500Name, boolean, java.util.List<net.corda.testing.node.User>, net.corda.testing.driver.VerifierType, net.corda.testing.node.ClusterSpec)
```

Comment (anthonykeenan) [reply testing/node-driver/src/main/kotlin/net/corda/testing/node/NotarySpec.kt]: A good example where we've had to do something like this before is in `LinearStateQueryCriteria`

File: docs/source/cordapp-build-systems.rst (hunk 14 above)
Comment (dimosr) [reply docs/source/cordapp-build-systems.rst]: Based on yesterday's discussion, I understood that you were good to go. But, no problem, I will do it now.

File: common/classloading/build.gradle (hunk 6 above)
Comment (shamsasari) [reply common/classloading/build.gradle]: Agree. Please move them to `core.internal` if possible.

File: common/classloading/src/main/kotlin/net/corda/common/classloading/Utils.kt [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+ * @param clazz the class of the interface, which the classes - to be returned - must implement.
+ *
+ * @return instances of the identified classes.
+ * @throws IllegalArgumentException if the classes found do not have proper constructors.
+ *
+ * Note: In order to be instantiated, the associated classes must:
+ * - be non-abstract
+ * - either be a Kotlin object or have a constructor with no parameters (or only optional ones)
+ */
+fun <T: Any> loadClassesImplementing(classloader: ClassLoader, clazz: Class<T>): Set<T> {
```
Comment (shamsasari) [inline common/classloading/src/main/kotlin/net/corda/common/classloading/Utils.kt]: This is not returning classes but object instances

File: common/classloading/src/main/kotlin/net/corda/common/classloading/Utils.kt [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+    return ClassGraph().addClassLoader(classloader)
+            .enableAllInfo()
+            .scan()
+            .getClassesImplementing(clazz.name)
+            .filterNot { it.isAbstract }
+            .mapNotNull { classloader.loadClass(it.name).asSubclass(clazz) }
```
Comment (shamsasari) [inline common/classloading/src/main/kotlin/net/corda/common/classloading/Utils.kt]: The expression is not nullable so `map` can be used instead

File: common/classloading/src/main/kotlin/net/corda/common/classloading/Utils.kt [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+            .map { it.kotlin.objectInstance ?: it.kotlin.createInstance() }
```
Comment (shamsasari) [inline common/classloading/src/main/kotlin/net/corda/common/classloading/Utils.kt]: Use `kotlinObjectInstance` instead

File: client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt [hunk 19]
```diff
@@ -252,8 +257,9 @@ class CordaRPCClient private constructor(
 ) {
     @JvmOverloads
     constructor(hostAndPort: NetworkHostAndPort,
-                configuration: CordaRPCClientConfiguration = CordaRPCClientConfiguration.DEFAULT)
-            : this(hostAndPort, configuration, null)
+                configuration: CordaRPCClientConfiguration = CordaRPCClientConfiguration.DEFAULT,
+                classLoader: ClassLoader? = null)
```
Comment (shamsasari) [inline client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt]: This is an kotlin API break

File: client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt [hunk 20]
```diff
@@ -240,6 +244,7 @@ open class CordaRPCClientConfiguration @JvmOverloads constructor(
  * @param configuration An optional configuration used to tweak client behaviour.
  * @param sslConfiguration An optional [ClientRpcSslOptions] used to enable secure communication with the server.
  * @param haAddressPool A list of [NetworkHostAndPort] representing the addresses of servers in HA mode.
+ * @param classLoader a classloader, which will be used (if provided) to discover available [SerializationCustomSerializer]s and [SerializationWhitelist]s
```
Comment (shamsasari) [inline client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt]: Can you also mention in the KDocs above when passing a separate classloader is required.

File: docs/source/cordapp-build-systems.rst (hunk 14 above)
Comment (tudor-malene) [reply docs/source/cordapp-build-systems.rst]: Thank you Dimos

**Comment by GavT** [comment]
Choosing to force a squash and merge. All integration tests passed successfully but reviewer has not given a green tick - this is understood and a follow-up PR is required which is not a show stopper for cutting C4 RC 01 and testing can start.
